pkg/guestagent/kubernetesservice/kubernetesservice.go

```diff
@@ -38,96 +38,161 @@ type Entry struct {
-				continue
-			}
+func canGetServices(ctx context.Context, kubectl, kubeconfig string) error {
+	cmd := exec.CommandContext(ctx, kubectl, "auth", "can-i", "get", "service")
+	if kubeconfig != "" {
+		cmd.Env = append(os.Environ(), "KUBECONFIG="+kubeconfig)
+	}
+	var stdout, stderr bytes.Buffer
+	cmd.Stdout = &stdout
+	cmd.Stderr = &stderr
+	if err := cmd.Run(); err != nil {
+		return fmt.Errorf("failed to run %v: %w; stdout=%q, stderr=%q", cmd.Args, err, stdout.String(), stderr.String())
+	}
+	if strings.TrimSpace(stdout.String()) != "yes" {
+		return fmt.Errorf("failed to run %v: expected \"yes\", got %q", cmd.Args, stdout.String())
+	}
+	return nil
+}
 
-			return nil, fmt.Errorf("stat kubeconfig %s failed: %w", kubeconfig, err)
-		}
+func (s *ServiceWatcher) startAndStreamKubectl(cmd *exec.Cmd) error {
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return err
+	}
+
+	var stderr bytes.Buffer
+	cmd.Stderr = &stderr
 
-		restConfig, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
-		if err != nil {
-			return nil, fmt.Errorf("build kubeconfig from %s failed: %w", kubeconfig, err)
+	if err := cmd.Start(); err != nil {
+		return fmt.Errorf("failed to run %v: %w; stderr=%q", cmd.Args, err, stderr.String())
+	}
+
+	readErr := s.readKubectlStream(stdout)
+	waitErr := cmd.Wait()
+	if waitErr != nil {
+		waitErr = fmt.Errorf("failed to run %v: %w; stderr=%q", cmd.Args, waitErr, stderr.String())
+	}
+	return errors.Join(readErr, waitErr)
+}
+
+// readKubectlStream reads kubectl JSON watch events from r and updates the internal
+// services map. The stream is newline-delimited JSON objects representing "WatchEvent".
+func (s *ServiceWatcher) readKubectlStream(r io.Reader) error {
+	scanner := bufio.NewScanner(r)
+	// increase buffer in case of large JSON objects
+	const maxBuf = 10 * 1024 * 1024
+	buf := make([]byte, 0, 64*1024)
+	scanner.Buffer(buf, maxBuf)
+
+	for scanner.Scan() {
+		line := scanner.Bytes()
+		line = bytes.TrimSpace(line)
+		if len(line) == 0 {
+			continue
 		}
-		u, err := url.Parse(restConfig.Host)
-		if err != nil {
-			return nil, fmt.Errorf("parse kubeconfig host %s failed: %w", restConfig.Host, err)
+
+		var ev struct {
+			Type   watch.EventType `json:"type"`
+			Object json.RawMessage `json:"object"`
 		}
-		if u.Hostname() != "127.0.0.1" { // might need to support IPv6
-			// ensures the kubeconfig points to local k8s
-			continue
+		if err := json.Unmarshal(line, &ev); err != nil {
+			return fmt.Errorf("failed to unmarshal line %q: %w", string(line), err)
 		}
 
-		kubeClient, err := kubernetes.NewForConfig(restConfig)
-		if err != nil {
-			return nil, err
+		var svc corev1.Service
+		if err := json.Unmarshal(ev.Object, &svc); err != nil {
+			return fmt.Errorf("failed to unmarshal service object: %w (line=%q)", err, line)
 		}
 
-		return kubeClient, nil
+		key := svc.Namespace + "/" + svc.Name
+		s.rwMutex.Lock()
+		switch ev.Type {
+		case watch.Added, watch.Modified:
+			s.services[key] = &svc
```

Comment:
Doe we need to keep all services in memory? We can extract the relevant details here and minimize memory usage.

Reply:
Fixed